src/App/Fossa/ManualDeps.hs

```diff
@@ -27,36 +28,51 @@ import Data.List.NonEmpty qualified as NE
 import Data.String.Conversion (toText)
 import Data.Text (Text, unpack)
 import DepTypes (DepType (..))
-import Effect.ReadFS (ReadFS, doesFileExist, readContentsYaml)
+import Effect.ReadFS (ReadFS, doesFileExist, readContentsJson, readContentsYaml)
 import Path
 import Srclib.Converter (depTypeToFetcher)
 import Srclib.Types (AdditionalDepData (..), Locator (..), SourceUnit (..), SourceUnitBuild (..), SourceUnitDependency (SourceUnitDependency), SourceUserDefDep (..))
 
-analyzeFossaDepsYaml :: (Has Diagnostics sig m, Has ReadFS sig m) => Path Abs Dir -> m (Maybe SourceUnit)
-analyzeFossaDepsYaml root = do
+data FoundDepsFile
+  = ManualYaml (Path Abs File)
+  | ManualJSON (Path Abs File)
+
+analyzeFossaDepsFile :: (Has Diagnostics sig m, Has ReadFS sig m) => Path Abs Dir -> m (Maybe SourceUnit)
+analyzeFossaDepsFile root = do
   maybeDepsFile <- findFossaDepsFile root
   case maybeDepsFile of
     Nothing -> pure Nothing
     -- If the file exists and we have no SourceUnit to report, that's a failure
     Just depsFile -> do
-      yamldeps <- context "Reading fossa-deps file" $ readContentsYaml depsFile
-      context "Converting fossa-deps to partial API payload" $ Just <$> toSourceUnit root yamldeps
+      manualDeps <- context "Reading fossa-deps file" $ readFoundDeps depsFile
+      context "Converting fossa-deps to partial API payload" $ Just <$> toSourceUnit root manualDeps
+
+readFoundDeps :: (Has Diagnostics sig m, Has ReadFS sig m) => FoundDepsFile -> m ManualDependencies
+readFoundDeps (ManualJSON path) = readContentsJson path
+readFoundDeps (ManualYaml path) = readContentsYaml path
 
-findFossaDepsFile :: (Has Diagnostics sig m, Has ReadFS sig m) => Path Abs Dir -> m (Maybe (Path Abs File))
+findFossaDepsFile :: (Has Diagnostics sig m, Has ReadFS sig m) => Path Abs Dir -> m (Maybe FoundDepsFile)
 findFossaDepsFile root = do
   let ymlFile = root </> $(mkRelFile "fossa-deps.yml")
       yamlFile = root </> $(mkRelFile "fossa-deps.yaml")
+      jsonFile = root </> $(mkRelFile "fossa-deps.json")
+      multipleFound = fatalText "Found multiple fossa-deps files.  Only one of ('.json', '.yml', and '.yaml') extensions are allowed"
   ymlExists <- doesFileExist ymlFile
   yamlExists <- doesFileExist yamlFile
-  case (ymlExists, yamlExists) of
-    (True, True) -> fatalText "Found '.yml' and '.yaml' files when searching for fossa-deps, only one is permitted if present."
-    (True, False) -> pure $ Just ymlFile
-    (False, True) -> pure $ Just yamlFile
-    (False, False) -> pure Nothing
-
-toSourceUnit :: Has Diagnostics sig m => Path Abs Dir -> YamlDependencies -> m SourceUnit
-toSourceUnit root yamldeps@YamlDependencies{..} = do
-  when (hasNoDeps yamldeps) $ fatalText "No dependencies found in fossa-deps file"
+  jsonExists <- doesFileExist jsonFile
+  case (ymlExists, yamlExists, jsonExists) of
+    -- Allow 0 or 1 files, not multiple
+    (True, True, _) -> multipleFound
```

Comment:
I think this is pretty clean.  The problem is that it's too verbose.  I've looked for stuff like this, but it would require quite a large abstraction to become much more readable.  I just don't think it's worth it.  I know it's a lot of text that feels removable, but it is crystal clear.

Reply:
There is some magic we can do here if we consider a new datatype, `AtMostOne a`, that represents finding at most one `a` -- and give it the semigroup/monoid instances we want:

```hs
data AtMostOne a = None | One a | MoreThanOne

instance Semigroup (AtMostOne a) where
  None <> a = a
  One a <> None = One a -- combining One with None produces One
  One _ <> _ = MoreThanOne -- combining One with anything else produces MoreThanOne
  MoreThanOne <> _ = MoreThanOne

instance Monoid (AtMostOne a) where
  mempty = None
```

and we can use this to our advantage, along with the existing functions:
```hs
-- type signature is self-explanatory
filterM :: Applicative m => (a -> m Bool) -> [a] -> m [a]

-- apply the function to turn each element in the list into something we can smash together with (<>), and smash everything together
foldMap :: (Traversable t, Monoid m) => (a -> m) -> t a -> m
-- or monomorphized:
foldMap :: (a -> AtMostOne a) -> [a] -> AtMostOne a
```

```hs
depsFilePath :: FoundDepsFile -> Path Abs File
depsFilePath (ManualYaml f) = f
depsFilePath (ManualJSON f) = f

findFossaDepsFile :: (Has Diagnostics sig m, Has ReadFS sig m) => Path Abs Dir -> m (Maybe FoundDepsFile)
findFossaDepsFile root = do
  let files =
        [ ManualYaml (root </> $(mkRelFile "fossa-deps.yml"))
        , ManualYaml (root </> $(mkRelFile "fossa-deps.yaml"))
        , ManualJSON (root </> $(mkRelFile "fossa-deps.json"))
        ]
  found <- filterM (doesFileExist . depsFilePath) files
  let found' = foldMap One found
  case found' of
    None -> pure Nothing
    One a -> pure (Just a)
    MoreThanOne -> fatalText "Found multiple fossa-deps files.  Only one of ('.json', '.yml', and '.yaml') extensions are allowed"
```

---

But it turns out we can build a less-magical version too:

```hs
depsFilePath :: FoundDepsFile -> Path Abs File
depsFilePath (ManualYaml f) = f
depsFilePath (ManualJSON f) = f

findFossaDepsFile :: (Has Diagnostics sig m, Has ReadFS sig m) => Path Abs Dir -> m (Maybe FoundDepsFile)
findFossaDepsFile root = do
  let files =
        [ ManualYaml (root </> $(mkRelFile "fossa-deps.yml"))
        , ManualYaml (root </> $(mkRelFile "fossa-deps.yaml"))
        , ManualJSON (root </> $(mkRelFile "fossa-deps.json"))
        ]
  found <- filterM (doesFileExist . depsFilePath) files
  case found of
    [] -> pure Nothing
    [a] -> pure (Just a)
    _ -> fatalText "Found multiple fossa-deps files.  Only one of ('.json', '.yml', and '.yaml') extensions are allowed"
```